### PR TITLE
Use appropriate connector instance based on protocol

### LINKED ISSUE #242: Can't access application after configure SSL
Hey folks, I'm working configuring Trino Gateway to use with our current Trino Cluster (running with HTTPS).
After a lot of tries, and some team member help, we are able to make the application working without configuration issues, here the relevant part of the config file:

```
requestRouter:
  ssl: true
  port: 9080
  name: trinoRouter
  historySize: 1000
  keystorePath: <path>
  keystorePass: <password>

server:
  applicationConnectors:
    - type: https
      port: 9081
      keyStorePath: <path>
      keyStorePassword: <password>
      useForwardedHeaders: true
  adminConnectors:
    - type: https
      port: 9082
      keyStorePath: <path>
      keyStorePassword: <password>
      useForwardedHeaders: true
```

Most of these configurations are from the [Dropwizard](https://www.dropwizard.io/en/stable/manual/configuration.html) docs, since we don't have detailed docs on how to deploy Trino Gateway with HTTPS.

After applying these configurations, the application starts, and was able to run healthychecks with Trino clusters, but the Web UI is not available.
Here an error log found:

```
INFO  [2024-02-08 21:03:44,350] io.trino.gateway.baseapp.BaseApp: op=register type=auth filter item=class io.dropwizard.auth.AuthFilter
ERROR [2024-02-08 21:03:44,373] io.trino.gateway.baseapp.BaseApp: Error loading managed app
! java.lang.IllegalStateException: null
! at java.base/java.util.OptionalInt.orElseThrow(OptionalInt.java:273)
! at io.trino.gateway.baseapp.AppModule.getApplicationPort(AppModule.java:37)
! at io.trino.gateway.ha.module.HaGatewayProviderModule.getProxyHandler(HaGatewayProviderModule.java:160)
! at io.trino.gateway.ha.module.HaGatewayProviderModule.provideGateway(HaGatewayProviderModule.java:206)
! at io.trino.gateway.ha.module.HaGatewayProviderModule$$FastClassByGuice$$48db3.GUICE$TRAMPOLINE(<generated>)
! at io.trino.gateway.ha.module.HaGatewayProviderModule$$FastClassByGuice$$48db3.apply(<generated>)
! at com.google.inject.internal.ProviderMethod$FastClassProviderMethod.doProvision(ProviderMethod.java:260)
! at com.google.inject.internal.ProviderMethod.doProvision(ProviderMethod.java:171)
! at com.google.inject.internal.InternalProviderInstanceBindingImpl$CyclicFactory.provision(InternalProviderInstanceBindingImpl.java:185)
! at com.google.inject.internal.InternalProviderInstanceBindingImpl$CyclicFactory.get(InternalProviderInstanceBindingImpl.java:162)
! at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
! at com.google.inject.internal.SingletonScope$1.get(SingletonScope.java:169)
! at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:45)
! at com.google.inject.internal.SingleFieldInjector.inject(SingleFieldInjector.java:50)
! at com.google.inject.internal.MembersInjectorImpl.injectMembers(MembersInjectorImpl.java:146)
! at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:124)
! at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:91)
! at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:300)
! at com.google.inject.internal.InjectorImpl$1.get(InjectorImpl.java:1148)
! ... 13 common frames omitted
! Causing: com.google.inject.ProvisionException: Unable to provision, see the following errors:
!
! 1) [Guice/ErrorInCustomProvider]: IllegalStateException
!   at HaGatewayProviderModule.provideGateway(HaGatewayProviderModule.java:187)
!   at GatewayManagedApp.gateway(GatewayManagedApp.java:7)
!       \_ for field gateway
!   while locating GatewayManagedApp
!
! Learn more:
!   https://github.com/google/guice/wiki/ERROR_IN_CUSTOM_PROVIDER
!
! 1 error
!
! ======================
! Full classname legend:
! ======================
! GatewayManagedApp:       "io.trino.gateway.ha.GatewayManagedApp"
! HaGatewayProviderModule: "io.trino.gateway.ha.module.HaGatewayProviderModule"
! ========================
! End of classname legend:
! ========================
!
! at com.google.inject.internal.InternalProvisionException.toProvisionException(InternalProvisionException.java:251)
! at com.google.inject.internal.InjectorImpl$1.get(InjectorImpl.java:1151)
! at com.google.inject.internal.InjectorImpl.getInstance(InjectorImpl.java:1186)
! at io.trino.gateway.baseapp.BaseApp.lambda$addManagedApps$0(BaseApp.java:189)
! at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
! at io.trino.gateway.baseapp.BaseApp.addManagedApps(BaseApp.java:184)
! at io.trino.gateway.baseapp.BaseApp.registerWithInjector(BaseApp.java:135)
! at io.trino.gateway.baseapp.BaseApp.configureGuice(BaseApp.java:125)
! at io.trino.gateway.baseapp.BaseApp.run(BaseApp.java:102)
! at io.trino.gateway.baseapp.BaseApp.run(BaseApp.java:48)
! at io.dropwizard.core.cli.EnvironmentCommand.run(EnvironmentCommand.java:66)
! at io.dropwizard.core.cli.ConfiguredCommand.run(ConfiguredCommand.java:98)
! at io.dropwizard.core.cli.Cli.run(Cli.java:78)
! at io.dropwizard.core.Application.run(Application.java:94)
! at io.trino.gateway.ha.HaGatewayLauncher.main(HaGatewayLauncher.java:25)
```

We don't know if that is a configuration issue or a bug in the application, so any help on this will be appreciated.

### PR DESCRIPTION
**Pull request by prakhar10**
This is to fix #242

### CODE REVIEW

**Comment by prakhar10** [comment]
@mosabua @ebyhr I'm not really sure how such cases are being handled in Trino repo, but i can take a look. If there are any suggestions please let me know.

File: gateway-ha/src/test/java/io/trino/gateway/baseapp/TestAppModule.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import org.mockito.Mockito;
+
+import java.util.Collections;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.Mockito.when;
+
+public class TestAppModule
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/baseapp/TestAppModule.java:29]: This test passes even before changing AppModule. It means the change isn't covered.

**Comment by mosabua** [comment]
Can you fix the test as requested @prakhar10

**Comment by prakhar10** [comment]
> Can you fix the test as requested @prakhar10

Sorry for the delay, got busy with something else. I'll fix this in a couple of days.

File: gateway-ha/src/test/java/io/trino/gateway/baseapp/TestAppModule.java (hunk 1 above)
Comment (mosabua) [reply gateway-ha/src/test/java/io/trino/gateway/baseapp/TestAppModule.java:29]: Any chance you can verify @ebyhr .. @prakhar10 pushed an update to the test .. I wont have time to verify

Comment (ebyhr) [reply gateway-ha/src/test/java/io/trino/gateway/baseapp/TestAppModule.java:29]: I confirmed the test still doesn't cover the change.

Comment (prakhar10) [reply gateway-ha/src/test/java/io/trino/gateway/baseapp/TestAppModule.java:30]: Sorry earlier changes didnt get pushed for some reason, please check now.

**Comment by prakhar10** [comment]
> Should we just use the changes from #294 with the test from this PR and improve in this PR as next step? Could you do that @prakhar10 ?

Sure @mosabua , but I believe we started off with a similar change and with reviews we've come to this code that I have currently in this PR.

But if it helps and if everyone agrees I can bring in the code changes from that PR into this one and we can take it forward.

**Comment by mosabua** [comment]
> > Should we just use the changes from #294 with the test from this PR and improve in this PR as next step? Could you do that @prakhar10 ?
> 
> Sure @mosabua , but I believe we started off with a similar change and with reviews we've come to this code that I have currently in this PR.
> 
> But if it helps and if everyone agrees I can bring in the code changes from that PR into this one and we can take it forward.

Ideally we just address the outstanding comments from @ebyhr ... and ensure the test works as expected. I really think we need to get this merged for the next release.

**Comment by ebyhr** [comment]
By the way, test isn't mandatory in my opinion. This is pretty small change and will be removed when migrating to Airlift.

**Comment by ebyhr** [comment]
Could you rebase on master to resolve conflicts?

**Comment by prakhar10** [comment]
> Could you rebase on main to resolve conflicts?

Looks like this file does not exist anymore in the main branch, do we need this PR then @ebyhr ?

**Comment by ebyhr** [comment]
@prakhar10 Please take a look at HaGatewayProviderModule:
https://github.com/trinodb/trino-gateway/blob/3c202992dd84c31829421bb1e9c791c18ccbe787/gateway-ha/src/main/java/io/trino/gateway/ha/module/HaGatewayProviderModule.java#L168-L183

**Comment by mosabua** [comment]
@prakhar10 are you going to move this change over to the new module? Even potentially without test .. just so we can get this merged and maybe ship as part of the 8 release?

File: gateway-ha/src/main/java/io/trino/gateway/baseapp/AppModule.java [hunk 2]
```diff
@@ -43,9 +44,12 @@ protected int getApplicationPort()
                         .map(SimpleServerFactory::getConnector);
 
         return connectors
-                .filter(connector -> connector.getClass().isAssignableFrom(HttpConnectorFactory.class))
-                .map(connector -> (HttpConnectorFactory) connector)
-                .mapToInt(HttpConnectorFactory::getPort)
+                .mapToInt(connector -> {
+                    if (connector instanceof HttpsConnectorFactory httpsConnectorFactory) {
+                        return httpsConnectorFactory.getPort();
```
[the remark this answers is not in the document]
Comment (mosabua) [reply gateway-ha/src/main/java/io/trino/gateway/baseapp/AppModule.java:49]: I looked at porting this to the new class now and also this code. Honestly I share the questions from @ebyhr and I am not sure at all why this change would do anything. I looked the the source of the two factory classes as well and the https one just extends the http one and there is no modification or change or so for the port. 

So if this change really resolves the issues.. I dont know why. Thoughts @prakhar10 ?

**Comment by willmostly** [comment]
Hi @prakhar10 , do you need any assistance with this?

**Comment by mosabua** [comment]
We discussed this in the recent a dev sync.. the reason this issue occurs and the fix partially fixes is that we use two separate HTTP ports for UI and API. However with the airlift refactor we will drop this usage down to just using one port (like Trino) and therefore this fix is not needed.

**Pull request closed without merging**